### PR TITLE
Combobox: Measure text of longest label for dropdown width

### LINKED ISSUE #92586: New Select: Use measureText for more accurate width estimation
See discussion about previous issues with "guesstimating" width

https://github.com/grafana/grafana/pull/86444#discussion_r1569033961

- Use measureText on the longest option every hundred results.
- Take padding into account

This should also remove the dynamic sizing functionality, and only find the longest option within the first 100 or so options.

### PR DESCRIPTION
**Pull request by joshhunt**
Fixes https://github.com/grafana/grafana/issues/92586

Changes combobox to use `measureText()` to size the dropdown menu more accurately. We assume (but have not measured) that `measureText()` is slow enough to not measure every item, so for optimisation we only measure the text of the longest item in the first 1000 options.

Tbh, we can probably look at many more items. on my (fast) computer, it takes me 600ms to look over 8 billion items. we can probably find a middleground in there somewhere.

By measuring only one item, if there a bunch of labels with similar length but different width, we may get truncated labels. Would be nice to _measure_ measureText() to see if it's worth looking at more items.

### CODE REVIEW

File: packages/grafana-ui/src/components/Combobox/getComboboxStyles.ts [hunk 1]
```diff
@@ -53,6 +57,9 @@ export const getComboboxStyles = (theme: GrafanaTheme2) => {
       label: 'grafana-select-option-label',
       textOverflow: 'ellipsis',
       overflow: 'hidden',
+      fontSize: MENU_ITEM_FONT_SIZE,
+      fontWeight: 500,
+      letterSpacing: 0, // pr todo: text in grafana has a slightly different letter spacing, which causes measureText() to be ~5% off
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Combobox/getComboboxStyles.ts:62]: The complication of `measureText()` is that we need to make sure the text is styled _exactly_ the same across that canvas and in the page.

**Comment by leeoniya** [comment]
> Tbh, we can probably look at many more items. on my (fast) computer, it takes me 600ms to look over 8 billion items. we can probably find a middleground in there somewhere.

you might be hitting the cache:

https://github.com/grafana/grafana/blob/86faeae6d299dc9b2d68fc9096cac7136cf0a0a1/packages/grafana-ui/src/utils/measureText.ts#L22

**Comment by joshhunt** [comment]
sorry @leeoniya, i was unclear! Outside of grafana i did a benchmark of _just_ doing `text.length` on a bunch of strings - I was curious whether there was any significant difference between things like Math.max vs a ternery (no, not really) for finding the longest string.

**Comment by leeoniya** [comment]
ah okay.

i've def seen a difference between `value = Math.max(value, thisValue)` and `if (thisValue > value) value = thisValue` in hot loops.

File: packages/grafana-ui/src/components/Combobox/useComboboxFloat.ts [hunk 2]
```diff
@@ -48,46 +48,25 @@ export const useComboboxFloat = (
...
-      const calculationStart = Math.max(0, startVisibleIndex - INDEX_WIDTH_CALCULATION);
-
-      for (let i = calculationStart; i < endVisibleIndex; i++) {
-        maxLength = Math.max(maxLength, items[i].label.length);
-      }
-
-      setPopoverWidth(maxLength * WIDTH_MULTIPLIER);
-    }
-  }, [items, range, setPopoverWidth]);
+    // pr todo: ideally we want to remove these magic numbers from here, or derive them from the styles
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/useComboboxFloat.ts:63]: Can we do some kind of measurement here?

Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/useComboboxFloat.ts:63]: what?

Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/useComboboxFloat.ts:63]: @tskarhed what?

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/useComboboxFloat.ts:63]: Well it is either taking it from the styles (which might be preferred) or doing an actual measuring of the element (without the text)?

File: packages/grafana-ui/src/components/Combobox/Combobox.story.tsx [hunk 3]
```diff
@@ -107,6 +108,123 @@ const ManyOptionsStory: StoryFn<PropsAndCustomArgs> = ({ numberOfOptions, ...arg
...
+        />
+      </Field>
+
+      <Divider />
+
+      <Field label="Combobox with auto width, minWidth 15">
+        <Combobox
+          id="combobox-auto-size"
+          width="auto"
+          minWidth={15}
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.story.tsx:175]: Is there any reason the search doesn't work here?

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.story.tsx:175]: And this should probably be wider
image

Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/Combobox.story.tsx:175]: oh good catch. I tested on windows chrome with fixed scroll bars and it worked fine, but not on mac with fixed scrollbars. ill give it another try :)

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.story.tsx:175]: The filtering is related to the auto width changes. I must have missed that.

**Review by joshhunt — COMMENTED** [review]
Safari on macOS will clip the text, with both floating or fixed scrollbars. I guess the text is being measured incorrectly in Safari? 🤔 


Screenshot 2024-10-09 at 12 14 08 PM

**Comment by tskarhed** [comment]
Oops, sorry

**Comment by tskarhed** [comment]
> Safari on macOS will clip the text, with both floating or fixed scrollbars. I guess the text is being measured incorrectly in Safari? 🤔
> 

I am using Brave on MacOS, so basically Chromium.

**Review by tskarhed — APPROVED** [review]
Clean up the comments and we're good to merge

**Pull request merged**